### PR TITLE
Implement api view for accessing archived stats data.

### LINKED ISSUE #1414: Create a view or API to allow accessing stats from the disk
To support #1395, let's make an API that allows the users to access the stats that we archived. It would probably have to return a list of stats that are available. Then allow a user to click to download the stats.

The permission around this should be the same as viewing stats in the graphs. Basically - if you've made them public anyone could get them. If not, only those that have the permission.

### PR DESCRIPTION
**Pull request by EnTeQuAk**
Fixes #1414

### CODE REVIEW

File: apps/stats/views.py [hunk 1]
```diff
@@ -699,3 +705,73 @@ def render_json(request, addon, stats):
...
+        """Fetches an addon by `id` or `slug`"""
+        qset = Addon.with_unlisted.all()
+
+        if not addon_id.isdigit():
+            addon = get_object_or_404(qset, slug=addon_id)
+        else:
+            addon = get_object_or_404(qset, guid=addon_id)
+
+        # If the addon is unlisted it needs either an owner/viewer/dev/support,
+        # or an unlisted addon reviewer.
```
Comment (magopian) [inline apps/stats/views.py]: Unlisted add-ons don't have stats at all, so we don't need this check, just return a 404 in any case.

File: apps/stats/views.py [hunk 2]
```diff
@@ -699,3 +705,73 @@ def render_json(request, addon, stats):
     fudge_headers(response, stats)
     json.dump(stats, response, cls=DecimalJSONEncoder)
     return response
+
+
+class ArchiveMixin(object):
+    """Provides common helper methods for all archive views."""
+    def get_addon(self, request, addon_id):
+        """Fetches an addon by `id` or `slug`"""
```
Comment (magopian) [inline apps/stats/views.py]: I'm not sure we want both, do we? Let's just keep it simple if possible and only support one.

File: apps/stats/views.py [hunk 3]
```diff
@@ -699,3 +705,73 @@ def render_json(request, addon, stats):
...
+            id=addon.id, year=year, month=month)
+
+        try:
+            files = storage.listdir(path)
+        except OSError:
+            return Response({
+                'error': 'No archived data for addon "%s" found.' % addon.slug,
+            }, status=status.HTTP_404_NOT_FOUND)
+
+        for file in files[1]:
```
Comment (magopian) [inline apps/stats/views.py]: Let's not use a builtin for a var name, could be `file_` instead

File: apps/stats/views.py [hunk 4]
```diff
@@ -699,3 +705,73 @@ def render_json(request, addon, stats):
...
+            file_name, _ = os.path.splitext(file)
+            parts = file_name.split('_')
```
Comment (magopian) [inline apps/stats/views.py]: use strptime instead?

File: apps/stats/views.py [hunk 5]
```diff
@@ -699,3 +705,73 @@ def render_json(request, addon, stats):
...
+
+        # If the addon is unlisted it needs either an owner/viewer/dev/support,
+        # or an unlisted addon reviewer.
+        owner_or_reviewer = owner_or_unlisted_reviewer(request, addon)
+        if not (addon.is_listed or owner_or_reviewer):
+            raise http.Http404
+        return addon
+
+
+class ArchiveListView(JWTProtectedView, ArchiveMixin):
```
Comment (magopian) [inline apps/stats/views.py]: Don't we usually put mixins left-most? I doubt it'll cause any issues here, but I think it's good practice

File: apps/stats/views.py [hunk 6]
```diff
@@ -24,10 +28,11 @@
 import amo
 from amo.utils import DecimalJSONEncoder
 from access import acl
-from addons.decorators import addon_view_factory
+from addons.decorators import addon_view_factory, owner_or_unlisted_reviewer
```
Comment (magopian) [inline apps/stats/views.py]: This is now not needed anymore I believe

File: docs/topics/api/stats.rst [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+We currently archive the following data:
+
++---------------------+--------------------+---------------------------------------+
+| name                |     model_name     | description                           |
++=====================+====================+=======================================+
+| update counts       | updatecounts       | How many users updated this addon     |
++---------------------+--------------------+---------------------------------------+
+| download counts     | downloadcounts     | How many users downloaded this addon  |
++---------------------+--------------------+---------------------------------------+
+| theme update counts | themeupdatecount   | How many users downloaded this theme  |
```
Comment (magopian) [inline docs/topics/api/stats.rst]: How many users have this theme installed (same above for update count: "how many users have this addon installed")

File: docs/topics/api/stats.rst [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
++---------------------+--------------------+---------------------------------------+
+
+--------------------------
+List archived data monthly
```
Comment (magopian) [inline docs/topics/api/stats.rst]: This seems odd, do you mean "List monthly archived data" ?

File: docs/topics/api/stats.rst [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    :param month: The month you want to fetch, please make sure to use 2 char months, e.g 01 instead of 1.
+
+    **Response:**
+
+    .. code-block:: json
+
+            [
+                {
+                    "date": "2016-01-18",
+                    "addon_id": 3615,
```
Comment (magopian) [inline docs/topics/api/stats.rst]: Should it be the addon slug instead? It says "slug" below, but in the code I do see the id being returned.

File: docs/topics/api/stats.rst [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+    **Response:**
+
+    .. code-block:: json
+
+            [
+                {
+                    "date": "2016-01-18",
+                    "count": 123,
+                    "addon": 3615
```
Comment (magopian) [inline docs/topics/api/stats.rst]: Same here, should it be id or slug?

File: docs/topics/api/stats.rst [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+                }
+            ]
+
+    :>json date: The full date for the data.
+    :>json count: The actual statistics data.
+    :>json addon: The addon id, can be used to relate and group data.
```
Comment (magopian) [inline docs/topics/api/stats.rst]: same here, id or slug?

File: apps/stats/tests/test_views.py [hunk 12]
```diff
@@ -878,3 +882,104 @@ def test_report_view_xss(self):
...
+    def clean_up_files(self):
+        path = os.path.join(views.storage.location, '3615')
+        if os.path.isdir(path):
+            shutil.rmtree(path)
+
+    def get(self, url=None):
+        return self.client.get(url, HTTP_AUTHORIZATION=self.authorization())
+
+    def test_list_not_authenticated(self):
+        # Use self.client.put so that we don't add the authorization header.
```
Comment (magopian) [inline apps/stats/tests/test_views.py]: Is this comment still relevant? I don't see a `self.client.put`

File: apps/stats/tests/test_views.py [hunk 13]
```diff
@@ -878,3 +882,104 @@ def test_report_view_xss(self):
...
+        self.user = UserProfile.objects.create(
+            read_dev_agreement=datetime.datetime.now())
+        self.api_key = self.create_api_key(self.user, 'bar')
+        response = self.get(
+            reverse('stats.archive_list', kwargs={
+                'slug': 'a3615', 'year': '2016', 'month': '01'}))
+        assert response.status_code == 403
+
+    def test_list(self):
+        self.user = UserProfile.objects.create(
```
Comment (magopian) [inline apps/stats/tests/test_views.py]: You didn't want to use the `self.user` from the `setUp`?

File: apps/stats/tests/test_views.py [hunk 14]
```diff
@@ -878,3 +882,104 @@ def test_report_view_xss(self):
...
+            read_dev_agreement=datetime.datetime.now())
+        self.api_key = self.create_api_key(self.user, 'bar')
+        AddonUser.objects.create(user=self.user, addon=self.addon)
+
+        response = self.get(
+            reverse('stats.archive', kwargs={
+                'slug': 'a3615', 'year': '2016', 'month': '01',
+                'day': '18', 'model_name': 'themeupdatecount'}))
+        assert response.status_code == 404
```
Comment (magopian) [inline apps/stats/tests/test_views.py]: You're not testing a `get` when not auth or when not owner, are you?

File: docs/topics/api/stats.rst (hunk 9 above)
Comment (EnTeQuAk) [inline docs/topics/api/stats.rst]: Well, this is the data we have saved on disk, I updated the doc below.

**Comment by magopian** [comment]
Looks good! r+wc

Are you going to create another PR for the frontend? I know the issue is named "API", but the part about "allow a user to click to download" makes me think this is also needed/wanted: to have the list of available archives in the stats pages, so users can just click to download them.

File: docs/topics/api/stats.rst (hunk 10 above)
Comment (EnTeQuAk) [inline docs/topics/api/stats.rst]: It's the data we saved on disk

File: apps/stats/tests/test_views.py (hunk 13 above)
Comment (EnTeQuAk) [inline apps/stats/tests/test_views.py]: Nope, I need an actual user that signed the dev agreement, that's why I'm creating a new one.

**Comment by EnTeQuAk** [comment]
Yeah, the frontend part is another PR once the API is finalized

File: apps/stats/tests/test_views.py (hunk 13 above)
Comment (magopian) [inline apps/stats/tests/test_views.py]: My point is that
1/ you declare a self.user in the setUp that you don't use
2/ you create a new self.user the exact same way (I think?) in each test

Maybe you could change the code in the setUp to do everything you need (accept the dev agreement, create the AddonUser).

File: apps/stats/tests/test_views.py (hunk 14 above)
Comment (EnTeQuAk) [inline apps/stats/tests/test_views.py]: Not sure I understand what you mean, this tests if the data exists. The addon is owned by the user and the user is able to use the api auth, so I'm testing the specific 404-part.

File: apps/stats/tests/test_views.py (hunk 13 above)
Comment (EnTeQuAk) [inline apps/stats/tests/test_views.py]: `self.user` is being used by `APIAuthTestCase` and used as a default for non-authenticated tests. I'm creating the new user again every time to explicitly make sure I have an authorized user.

**Pull request merged**